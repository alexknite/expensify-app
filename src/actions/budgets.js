import database from '../firebase/firebase';

export const addBudget = (budget) => ({
  type: 'ADD_BUDGET',
  budget
});

export const startAddBudget = (budgetData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      amount = 0,
      startDate = 0,
      endDate = 0,
      note = '',
      category = '',
    } = budgetData;
    const budget = { amount, startDate, endDate, note, category };

    return database.ref(`users/${uid}/budgets`).push(budget).then((ref) => {
      dispatch(addBudget({
        id: ref.key,
        ...budget
      }));
    });
  };
};

export const removeBudget = ({ id } = {}) => ({
    type: 'REMOVE_BUDGET',
    id
});

export const startRemoveBudget = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/budgets/${id}`).remove().then(() => {
      dispatch(removeBudget({ id }));
    });
  };
};

export const editBudget = (id, updates) => ({
  type: 'EDIT_BUDGET',
  id,
  updates
});

export const startEditBudget = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/budgets/${id}`).update(updates).then(() => {
      dispatch(editBudget(id, updates));
    });
  };
};

export const setBudgets = (budgets) => ({
  type: 'SET_BUDGETS',
  budgets
});

export const startSetBudgets = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/budgets`).once('value').then((snapshot) => {
      const budgets = [];

      snapshot.forEach((childSnapshot) => {
        budgets.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

       dispatch(setBudgets(budgets));
    });
  };
};
