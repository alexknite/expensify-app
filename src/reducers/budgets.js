const budgetsReducerDefState = [];

export default (state = budgetsReducerDefState, action) => {
  switch (action.type) {
    case 'ADD_BUDGET':
      return [
        ...state,
        action.budget
      ];
    case 'REMOVE_BUDGET':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_BUDGET':
      return state.map((budget) => {
        if (budget.id === action.id) {
          return {
            ...budget,
            ...action.updates
          };
        } else {
          return budget;
        }
      });
    case 'SET_BUDGETS':
      return action.budgets;
    default:
      return state;
  }
};
