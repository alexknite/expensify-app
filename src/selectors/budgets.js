import moment from 'moment';

export default (budgets, { text, sortBy, startDate, endDate, category }) => {
  return budgets.filter((budget) => {
    const startDateMoment = moment(budget.startDate);
    const endDateMoment = moment(budget.endDate);
    const startDateMatch = startDate ? startDate.isSameOrBefore(startDateMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(endDateMoment, 'day') : true;
    const categoryMatch = budget.category.toLowerCase().includes(category.toLowerCase());

    return startDateMatch && endDateMatch && categoryMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};
