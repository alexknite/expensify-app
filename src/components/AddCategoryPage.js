import React from 'react';
import { connect } from 'react-redux';
import CategoriesForm from './CategoriesForm';
import { startAddCategory } from '../actions/categories';

export class AddCategoryPage extends React.Component {
  onSubmit = (category) => {
    this.props.startAddCategory(category);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Category</h1>
          </div>
        </div>
        <div className="content-container">
          <CategoriesForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category)),
});

export default connect(undefined, mapDispatchToProps)(AddCategoryPage);
