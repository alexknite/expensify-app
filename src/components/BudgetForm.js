import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { DateRangePicker } from 'react-dates';
import { startAddCategory } from '../actions/categories';

export class BudgetForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      category: props.budget ? props.budget.category : '',
      amount: props.budget ? (props.budget.amount / 100).toString() : '',
      startDate: props.budget ? moment(props.budget.startDate) : moment(),
      endDate: props.budget ? moment(props.budget.endDate) : moment().add(7, 'days'),
      note: props.budget ? props.budget.note : '',
      calendarFocused: null
    };
  };
  onCategoryChange = (e) => {
    const category = e.target.value;
    this.setState(() => ({ category }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.setState(() => ({ startDate, endDate }));
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onSubmit = async (e) => {
    e.preventDefault();
    const cleansedCategories = this.props.categories.map(({ name }) => name
      .replace(/[^A-Z0-9]+/ig, '')
      .toLowerCase());
    const cleansedCategory = this.state.category
      .replace(/[^A-Z0-9]+/ig, '')
      .toLowerCase();
    const categoryIndex = cleansedCategories.indexOf(cleansedCategory);

    if (categoryIndex > -1) {
      await this.setState(() => ({ category: this.props.categories[categoryIndex].name }));
    } else {
      this.props.startAddCategory({ name: this.state.category });
    }

    if (!this.state.amount || (!this.state.startDate && !this.state.endDate) ) {
      this.setState(() => ({ error: 'Please provide the start date, end date and the amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        category: this.state.category,
        amount: parseFloat(this.state.amount, 10) * 100,
        startDate: this.state.startDate.valueOf(),
        endDate: this.state.endDate.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Category"
          autoFocus
          value={this.state.category}
          onChange={(this.onCategoryChange)}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <DateRangePicker
          startDate={this.state.startDate}
          endDate={this.state.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
        />
        <textarea
          className="textarea"
          placeholder="Add a note for your budget (optional)"
          value={this.state.note}
          onChange={(this.onNoteChange)}
        >
        </textarea>
        <div>
          <button className="button">Save Budget</button>
        </div>
      </form>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    categories: state.categories
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startAddCategory: ({ name }) => dispatch(startAddCategory({ name }))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(BudgetForm);
