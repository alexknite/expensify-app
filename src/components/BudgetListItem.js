import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { selectBudget } from '../actions/budget';

const BudgetListItem = ({ id, amount, startDate, endDate, note, selectBudget, category }) => {
  const onSelect = () => {
    selectBudget({
      id,
      amount,
      startDate,
      endDate,
      note,
      category
    });
  };
  return (
    <div className="list-item">
      <div>
        <h2 className="list-item__sub-title">{category}</h2>
        <h3 className="list-item__title">{numeral(amount / 100).format(',0.00')}</h3>
        <p className="list-item__sub-title">{moment(startDate).format('MMMM Do, YYYY')} - {moment(endDate).format('MMMM Do, YYYY')}</p>
        <h3 className="list-item__sub-title">{note}</h3>
      </div>
      <div className="list-item__actions">
        <button className="button button--link" onClick={onSelect}>
          <img src="/images/eye.svg" alt="Select"/>
        </button>
        <Link to={`edit/budget/${id}`}>
          <button className="button button--link">
            <img src="/images/edit.svg" alt="Edit"/>
          </button>
        </Link>
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    selectBudget: (budget) => dispatch(selectBudget(budget))
  };
};

export default connect(undefined, mapDispatchToProps)(BudgetListItem);
