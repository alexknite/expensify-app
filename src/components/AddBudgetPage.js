import React from 'react';
import { connect } from 'react-redux';
import BudgetForm from './BudgetForm';
import { startAddBudget } from '../actions/budgets';
import { selectBudget } from '../actions/budget';

export class AddBudgetPage extends React.Component {
  onSubmit = (budget) => {
    this.props.startAddBudget(budget);
    this.props.selectBudget({});
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Budget</h1>
          </div>
        </div>
        <div className="content-container">
          <BudgetForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  startAddBudget: (budget) => dispatch(startAddBudget(budget)),
  selectBudget: (budget) => dispatch(startAddBudget(budget))
});

export default connect(undefined, mapDispatchToProps)(AddBudgetPage);
