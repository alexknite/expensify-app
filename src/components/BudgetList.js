import React from 'react';
import { connect } from 'react-redux';
import BudgetListItem from './BudgetListItem';
import selectBudgets from '../selectors/budgets';

export const BudgetList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Budgets</div>
      <div className="show-for-desktop">Budget</div>
    </div>
    <div className="list-body">
      {
        props.budgets.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No budgets available</span>
          </div>
        ) : (
          props.budgets.map((budget) => {
            return <BudgetListItem key={budget.id} {...budget} />
          })
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    budgets: selectBudgets(state.budgets, state.budgetsFilters)
  };
};

export default connect(mapStateToProps)(BudgetList);
