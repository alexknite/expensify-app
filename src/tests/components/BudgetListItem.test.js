import React from 'react';
import { shallow } from 'enzyme';
import { BudgetListItem } from '../../components/BudgetListItem';
import budgets from  '../fixtures/budgets';

test('should render BudgetListItem correctly', () => {
  const wrapper = shallow(<BudgetListItem {...budgets[0]} />);
  expect(wrapper).toMatchSnapshot();
});
