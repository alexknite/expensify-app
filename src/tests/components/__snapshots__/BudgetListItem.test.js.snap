// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BudgetListItem correctly 1`] = `
<div
  className="list-item"
>
  <div>
    <h2
      className="list-item__sub-title"
    >
      Food
    </h2>
    <h3
      className="list-item__title"
    >
      1.00
    </h3>
    <p
      className="list-item__sub-title"
    >
      December 31st, 1969
       - 
      January 7th, 1970
    </p>
    <h3
      className="list-item__sub-title"
    />
  </div>
  <div
    className="list-item__actions"
  >
    <button
      className="button button--link"
      onClick={[Function]}
    >
      <img
        alt="Select"
        src="/images/eye.svg"
      />
    </button>
    <Link
      replace={false}
      to="edit/budget/1"
    >
      <button
        className="button button--link"
      >
        <img
          alt="Edit"
          src="/images/edit.svg"
        />
      </button>
    </Link>
  </div>
</div>
`;
